File: backend/recipe-service/src/main/java/com/smartoutlet/recipe/entity/Recipe.java
package com.smartoutlet.recipe.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "recipes", indexes = {
    @Index(name = "idx_product_id", columnList = "product_id"),
    @Index(name = "idx_recipe_code", columnList = "recipe_code"),
    @Index(name = "idx_status", columnList = "status"),
    @Index(name = "idx_version", columnList = "product_id, version")
})
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Recipe {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @Column(name = "product_id", nullable = false)
    private Long productId;
    
    @Column(name = "recipe_code", nullable = false, unique = true, length = 100)
    private String recipeCode;
    
    @Column(name = "name", nullable = false, length = 200)
    private String name;
    
    @Column(name = "description", length = 1000)
    private String description;
    
    @Column(name = "version", nullable = false)
    @Builder.Default
    private Integer version = 1;
    
    @Column(name = "batch_size", precision = 10, scale = 3)
    private BigDecimal batchSize;
    
    @Enumerated(EnumType.STRING)
    @Column(name = "batch_unit", length = 20)
    private UnitOfMeasure batchUnit;
    
    @Column(name = "yield_percentage", precision = 5, scale = 2)
    @Builder.Default
    private BigDecimal yieldPercentage = BigDecimal.valueOf(100.0);
    
    @Column(name = "preparation_time_minutes")
    private Integer preparationTimeMinutes;
    
    @Column(name = "cooking_time_minutes")
    private Integer cookingTimeMinutes;
    
    @Column(name = "total_time_minutes")
    private Integer totalTimeMinutes;
    
    @Column(name = "difficulty_level")
    @Builder.Default
    private Integer difficultyLevel = 1; // 1-5 scale
    
    @Column(name = "serving_size", precision = 10, scale = 3)
    private BigDecimal servingSize;
    
    @Enumerated(EnumType.STRING)
    @Column(name = "serving_unit", length = 20)
    private UnitOfMeasure servingUnit;
    
    @Column(name = "allergen_info", length = 1000)
    private String allergenInfo;
    
    @Column(name = "nutritional_info", length = 2000)
    private String nutritionalInfo;
    
    @Column(name = "dietary_restrictions", length = 500)
    private String dietaryRestrictions;
    
    @Column(name = "cooking_instructions", length = 5000)
    private String cookingInstructions;
    
    @Column(name = "storage_instructions", length = 1000)
    private String storageInstructions;
    
    @Column(name = "shelf_life_hours")
    private Integer shelfLifeHours;
    
    @Column(name = "temperature_requirement", length = 100)
    private String temperatureRequirement;
    
    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 20)
    @Builder.Default
    private RecipeStatus status = RecipeStatus.DRAFT;
    
    @Column(name = "cost_per_batch", precision = 10, scale = 4)
    private BigDecimal costPerBatch;
    
    @Column(name = "cost_per_serving", precision = 10, scale = 4)
    private BigDecimal costPerServing;
    
    @Column(name = "labor_cost_per_batch", precision = 10, scale = 4)
    private BigDecimal laborCostPerBatch;
    
    @Column(name = "overhead_cost_per_batch", precision = 10, scale = 4)
    private BigDecimal overheadCostPerBatch;
    
    @Column(name = "total_cost_per_batch", precision = 10, scale = 4)
    private BigDecimal totalCostPerBatch;
    
    @Column(name = "created_by")
    private Long createdBy;
    
    @Column(name = "approved_by")
    private Long approvedBy;
    
    @Column(name = "approved_at")
    private LocalDateTime approvedAt;
    
    @Column(name = "image_url", length = 500)
    private String imageUrl;
    
    @Column(name = "video_url", length = 500)
    private String videoUrl;
    
    @Column(name = "tags", length = 500)
    private String tags;
    
    @Column(name = "notes", length = 2000)
    private String notes;
    
    @OneToMany(mappedBy = "recipe", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<RecipeIngredient> ingredients;
    
    @OneToMany(mappedBy = "recipe", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<RecipeStep> steps;
    
    @OneToMany(mappedBy = "recipe", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<ProductionBatch> productionBatches;
    
    @CreationTimestamp
    @Column(name = "created_at")
    private LocalDateTime createdAt;
    
    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
    
    // Business logic methods
    public BigDecimal calculateMaterialCost() {
        if (ingredients == null || ingredients.isEmpty()) {
            return BigDecimal.ZERO;
        }
        
        return ingredients.stream()
                .map(RecipeIngredient::getTotalCost)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }
    
    public BigDecimal calculateTotalCost() {
        BigDecimal materialCost = calculateMaterialCost();
        BigDecimal labor = laborCostPerBatch != null ? laborCostPerBatch : BigDecimal.ZERO;
        BigDecimal overhead = overheadCostPerBatch != null ? overheadCostPerBatch : BigDecimal.ZERO;
        
        return materialCost.add(labor).add(overhead);
    }
    
    public BigDecimal getCostPerUnit() {
        if (batchSize == null || batchSize.compareTo(BigDecimal.ZERO) == 0) {
            return BigDecimal.ZERO;
        }
        return calculateTotalCost().divide(batchSize, 4, java.math.RoundingMode.HALF_UP);
    }
    
    public boolean isActive() {
        return status == RecipeStatus.ACTIVE;
    }
    
    public boolean isApproved() {
        return status == RecipeStatus.ACTIVE || status == RecipeStatus.APPROVED;
    }
    
    public int getTotalTimeRequired() {
        int prep = preparationTimeMinutes != null ? preparationTimeMinutes : 0;
        int cook = cookingTimeMinutes != null ? cookingTimeMinutes : 0;
        return prep + cook;
    }
    
    public BigDecimal getEffectiveYield() {
        if (yieldPercentage == null) return BigDecimal.valueOf(100);
        return yieldPercentage.divide(BigDecimal.valueOf(100), 4, java.math.RoundingMode.HALF_UP);
    }
}

File: frontend/src/types/recipe.ts
// Recipe and Raw Materials Types

export interface RawMaterial {
  id: number;
  materialCode: string;
  name: string;
  description?: string;
  category: string;
  subcategory?: string;
  unitOfMeasure: UnitOfMeasure;
  baseUnitCost: number;
  currentStock: number;
  minStockLevel?: number;
  maxStockLevel?: number;
  reorderPoint?: number;
  reorderQuantity?: number;
  leadTimeDays: number;
  shelfLifeDays?: number;
  storageTemperatureMin?: number;
  storageTemperatureMax?: number;
  storageConditions?: string;
  primaryVendor?: VendorSummary;
  allergenInfo?: string;
  nutritionalInfo?: string;
  originCountry?: string;
  certifications?: string;
  wastagePercentage: number;
  status: MaterialStatus;
  lastPurchaseDate?: string;
  lastPurchaseCost?: number;
  averageCost?: number;
  imageUrl?: string;
  barcode?: string;
  internalNotes?: string;
  createdAt: string;
  updatedAt: string;
  
  // Calculated fields
  stockValue: number;
  isLowStock: boolean;
  needsReorder: boolean;
  isExpired: boolean;
  daysUntilReorderNeeded: number;
  stockStatusColor: string;
}

export interface Recipe {
  id: number;
  productId: number;
  productName?: string;
  recipeCode: string;
  name: string;
  description?: string;
  version: number;
  batchSize: number;
  batchUnit: UnitOfMeasure;
  yieldPercentage: number;
  preparationTimeMinutes?: number;
  cookingTimeMinutes?: number;
  totalTimeMinutes?: number;
  difficultyLevel: number;
  servingSize?: number;
  servingUnit?: UnitOfMeasure;
  allergenInfo?: string;
  nutritionalInfo?: string;
  dietaryRestrictions?: string;
  cookingInstructions?: string;
  storageInstructions?: string;
  shelfLifeHours?: number;
  temperatureRequirement?: string;
  status: RecipeStatus;
  costPerBatch?: number;
  costPerServing?: number;
  laborCostPerBatch?: number;
  overheadCostPerBatch?: number;
  totalCostPerBatch?: number;
  imageUrl?: string;
  videoUrl?: string;
  tags?: string;
  notes?: string;
  ingredients: RecipeIngredient[];
  steps: RecipeStep[];
  createdBy?: UserSummary;
  approvedBy?: UserSummary;
  approvedAt?: string;
  createdAt: string;
  updatedAt: string;
  
  // Calculated fields
  materialCost: number;
  costPerUnit: number;
  statusColor: string;
  isActive: boolean;
  isApproved: boolean;
}

export interface RecipeIngredient {
  id?: number;
  rawMaterial: RawMaterialSummary;
  quantity: number;
  unit: UnitOfMeasure;
  stepNumber: number;
  preparationMethod?: string;
  notes?: string;
  isOptional: boolean;
  substituteIngredients?: string;
  unitCost: number;
  totalCost: number;
  wastagePercentage?: number;
  effectiveQuantity: number;
  allergenContribution?: string;
  processingInstructions?: string;
  qualityStandards?: string;
  costPercentageOfRecipe: number;
}

export interface RecipeStep {
  id?: number;
  stepNumber: number;
  title: string;
  description: string;
  durationMinutes?: number;
  temperature?: string;
  equipment?: string;
  tips?: string;
  imageUrl?: string;
  isCritical: boolean;
}

export interface Vendor {
  id: number;
  vendorCode: string;
  name: string;
  legalName?: string;
  description?: string;
  vendorType: VendorType;
  contactPerson?: string;
  email?: string;
  phone?: string;
  mobile?: string;
  website?: string;
  addressLine1?: string;
  addressLine2?: string;
  city?: string;
  state?: string;
  postalCode?: string;
  country?: string;
  taxId?: string;
  businessLicense?: string;
  paymentTerms?: string;
  creditLimit?: number;
  currentBalance: number;
  discountPercentage?: number;
  leadTimeDays: number;
  minimumOrderValue?: number;
  deliveryCharges?: number;
  qualityRating?: number;
  deliveryRating?: number;
  serviceRating?: number;
  overallRating?: number;
  certifications?: string;
  specialties?: string;
  preferredDeliveryDays?: string;
  deliveryTimeFrom?: string;
  deliveryTimeTo?: string;
  status: VendorStatus;
  notes?: string;
  lastOrderDate?: string;
  totalOrders: number;
  totalOrderValue: number;
  materialsSupplied?: string[];
  createdAt: string;
  updatedAt: string;
}

// Summary/Reference types
export interface VendorSummary {
  id: number;
  vendorCode: string;
  name: string;
  contactPerson?: string;
  email?: string;
  phone?: string;
  leadTimeDays: number;
  qualityRating?: number;
  overallRating?: number;
}

export interface RawMaterialSummary {
  id: number;
  materialCode: string;
  name: string;
  category: string;
  unitOfMeasure: UnitOfMeasure;
  baseUnitCost: number;
  currentStock: number;
  imageUrl?: string;
}

export interface UserSummary {
  id: number;
  username: string;
  firstName?: string;
  lastName?: string;
  email?: string;
}

// Enums
export enum UnitOfMeasure {
  // Weight units
  KILOGRAM = 'KILOGRAM',
  GRAM = 'GRAM',
  POUND = 'POUND',
  OUNCE = 'OUNCE',
  
  // Volume units
  LITER = 'LITER',
  MILLILITER = 'MILLILITER',
  GALLON = 'GALLON',
  QUART = 'QUART',
  PINT = 'PINT',
  CUP = 'CUP',
  TABLESPOON = 'TABLESPOON',
  TEASPOON = 'TEASPOON',
  
  // Count units
  PIECE = 'PIECE',
  DOZEN = 'DOZEN',
  PACK = 'PACK',
  BOTTLE = 'BOTTLE',
  CAN = 'CAN',
  BOX = 'BOX',
  BAG = 'BAG',
  
  // Length units
  METER = 'METER',
  CENTIMETER = 'CENTIMETER',
  INCH = 'INCH',
  FOOT = 'FOOT'
}

export enum MaterialStatus {
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
  DISCONTINUED = 'DISCONTINUED',
  PENDING_APPROVAL = 'PENDING_APPROVAL',
  OUT_OF_STOCK = 'OUT_OF_STOCK',
  RECALLED = 'RECALLED'
}

export enum RecipeStatus {
  DRAFT = 'DRAFT',
  PENDING_APPROVAL = 'PENDING_APPROVAL',
  APPROVED = 'APPROVED',
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
  ARCHIVED = 'ARCHIVED',
  REJECTED = 'REJECTED'
}

export enum VendorType {
  SUPPLIER = 'SUPPLIER',
  MANUFACTURER = 'MANUFACTURER',
  DISTRIBUTOR = 'DISTRIBUTOR',
  WHOLESALER = 'WHOLESALER',
  FARMER = 'FARMER',
  IMPORTER = 'IMPORTER',
  LOCAL_PRODUCER = 'LOCAL_PRODUCER',
  ORGANIC_SUPPLIER = 'ORGANIC_SUPPLIER',
  SPECIALTY_SUPPLIER = 'SPECIALTY_SUPPLIER'
}

export enum VendorStatus {
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
  SUSPENDED = 'SUSPENDED',
  BLACKLISTED = 'BLACKLISTED',
  PENDING_APPROVAL = 'PENDING_APPROVAL',
  ARCHIVED = 'ARCHIVED'
}

// Request/Form types
export interface RawMaterialCreateRequest {
  materialCode: string;
  name: string;
  description?: string;
  category: string;
  subcategory?: string;
  unitOfMeasure: UnitOfMeasure;
  baseUnitCost: number;
  minStockLevel?: number;
  maxStockLevel?: number;
  reorderPoint?: number;
  reorderQuantity?: number;
  leadTimeDays?: number;
  shelfLifeDays?: number;
  storageTemperatureMin?: number;
  storageTemperatureMax?: number;
  storageConditions?: string;
  primaryVendorId?: number;
  allergenInfo?: string;
  nutritionalInfo?: string;
  originCountry?: string;
  certifications?: string;
  wastagePercentage?: number;
  imageUrl?: string;
  barcode?: string;
  internalNotes?: string;
}

export interface RecipeCreateRequest {
  productId: number;
  recipeCode: string;
  name: string;
  description?: string;
  batchSize: number;
  batchUnit: UnitOfMeasure;
  yieldPercentage?: number;
  preparationTimeMinutes?: number;
  cookingTimeMinutes?: number;
  difficultyLevel?: number;
  servingSize?: number;
  servingUnit?: UnitOfMeasure;
  allergenInfo?: string;
  nutritionalInfo?: string;
  dietaryRestrictions?: string;
  cookingInstructions?: string;
  storageInstructions?: string;
  shelfLifeHours?: number;
  temperatureRequirement?: string;
  laborCostPerBatch?: number;
  overheadCostPerBatch?: number;
  imageUrl?: string;
  videoUrl?: string;
  tags?: string;
  notes?: string;
  ingredients: RecipeIngredientRequest[];
  steps: RecipeStepRequest[];
}

export interface RecipeIngredientRequest {
  rawMaterialId: number;
  quantity: number;
  unit: UnitOfMeasure;
  stepNumber?: number;
  preparationMethod?: string;
  notes?: string;
  isOptional?: boolean;
  substituteIngredients?: string;
  wastagePercentage?: number;
  processingInstructions?: string;
  qualityStandards?: string;
}

export interface RecipeStepRequest {
  stepNumber: number;
  title: string;
  description: string;
  durationMinutes?: number;
  temperature?: string;
  equipment?: string;
  tips?: string;
  imageUrl?: string;
  isCritical?: boolean;
}

// Filter types
export interface RawMaterialFilters {
  category?: string;
  status?: MaterialStatus;
  vendorId?: number;
  search?: string;
  page?: number;
  size?: number;
  sortBy?: string;
  sortDir?: 'asc' | 'desc';
}

export interface RecipeFilters {
  productId?: number;
  status?: RecipeStatus;
  search?: string;
  page?: number;
  size?: number;
  sortBy?: string;
  sortDir?: 'asc' | 'desc';
}

// Analysis types
export interface MaterialForecast {
  productId: number;
  forecastPeriodDays: number;
  expectedSalesQuantity: number;
  materialRequirements: MaterialRequirement[];
  totalEstimatedCost: number;
  recommendedOrders: RecommendedOrder[];
}

export interface MaterialRequirement {
  materialId: number;
  materialName: string;
  requiredQuantity: number;
  unit: string;
}

export interface RecommendedOrder {
  vendorId: number;
  vendorName: string;
  materials: string[];
}

export interface VendorReorderPlan {
  vendorId: number;
  vendorName: string;
  planningHorizonDays: number;
  recommendedOrders: ReorderRecommendation[];
  totalEstimatedCost: number;
  minimumOrderValue: number;
  meetsMinimumOrder: boolean;
}

export interface ReorderRecommendation {
  materialId: number;
  materialName: string;
  currentStock: number;
  reorderPoint: number;
  recommendedQuantity: number;
  unit: string;
  estimatedCost: number;
  urgency: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  expectedDeliveryDate: string;
}

export interface RecipeCostAnalysis {
  recipeId: number;
  batchQuantity: number;
  materialCost: number;
  laborCost: number;
  overheadCost: number;
  totalCost: number;
  costPerUnit: number;
  profitMargin: number;
  suggestedSellingPrice: number;
}

// Unit of measure metadata
export interface UnitOfMeasureInfo {
  value: string;
  symbol: string;
  displayName: string;
  category: string;
}